### PR TITLE
[WIP] Extend cache contexts via code.

### LINKED ISSUE #185: Extend cache contexts via code
From Wim Leers: https://www.drupal.org/node/2662196#comment-10838744

> OT: the README in that branch/repo says: "Edit the required_cache_contexts to add domain-awareness, as indicated by the url.site parameter." — that should not even be necessary. See #18. Look at \Drupal\language\LanguageServiceProvider::alter() for an example.

### PR DESCRIPTION
**Pull request by zerolab**
Fixes #185
See https://www.drupal.org/node/2662196#comment-10838164 for context.

Still reliant on the patch from https://www.drupal.org/node/2662196 for this to work.

### CODE REVIEW

**Comment by zerolab** [comment]
After I submitted this PR, I discovered not all tests were run on Travis as it was only enabling domain.
Now fixed. There are a few tests that fail, yes run-tests.sh does not report an error code.

I have re-added the `after_script` checks. See https://travis-ci.org/zerolab/domain/jobs/116272631 vs https://travis-ci.org/agentrickard/domain/jobs/116269032

Oddly enough I cannot repeat the failures when run manually on my local build. Will have another look tomorrow. In the meantime, does the PR make sense?

**Comment by agentrickard** [comment]
I believe so, yes.

I'm also about to be on vacation for a week, sans computer.

**Comment by agentrickard** [comment]
Let's retest this without the core patch and against #209

**Comment by agentrickard** [comment]
#209 has been merged, so this is going to need a re-roll.

**Comment by zerolab** [comment]
Re-rolled, patch removed.
This PR enables the tests for all modules. Currently failing for domain_accces, domain_config, domain_source.

Will submit another PR a bit later that will make the Travis check fail properly. I guess once that is in, we will need to fix failing tests :)

File: domain_config/src/DomainConfigServiceProvider.php [hunk 1]
```diff
@@ -8,19 +8,34 @@
...
   public function alter(ContainerBuilder $container) {
     $definition = $container->getDefinition('router.route_provider');
     $definition->setClass(DomainRouteProvider::class);
+
+    if ($container->hasParameter('renderer.config')) {
+      $renderer_config = $container->getParameter('renderer.config');
+
+      if (!in_array('url.site', $renderer_config['required_cache_contexts'])) {
+        $renderer_config['required_cache_contexts'][] = 'url.site';
+        sort($renderer_config['required_cache_contexts']);
```
Comment (agentrickard) [inline domain_config/src/DomainConfigServiceProvider.php]: What is the purpose of this sort?

Comment (zerolab) [inline domain_config/src/DomainConfigServiceProvider.php]: Good catch. I believe that was a late night look at how the [Renderer]() class works with cache contexts. `$elements['#cache']['contexts'] = Cache::mergeContexts($elements['#cache']['contexts'], $required_cache_contexts);` does the sorting.

I believe it interpreted that wrongly. Had another look and it doesn't seem we need it. Removed and squashed the previous commits into one

**Comment by agentrickard** [comment]
I love automated testing so much right now.

**Comment by zerolab** [comment]
Me too!
Thank you for taking the time to review and merge.

**Pull request merged**